extension/oauth2clientauthextension/clientcredentialsconfig.go

```diff
@@ -0,0 +1,105 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package oauth2clientauthextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/oauth2clientauthextension"
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"go.uber.org/multierr"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+// clientCredentialsConfig is a clientcredentials.Config wrapper to allow
+// values read from files in the ClientID and ClientSecret fields.
+//
+// Values from files can be retrieved by populating the ClientIDFile or
+// the ClientSecretFile fields with the path to the file.
+//
+// Priority: File > Raw value
+//
+// Example - Retrieve secret from file:
+//
+//	cfg := clientCredentialsConfig{
+//		Config: clientcredentials.Config{
+//			ClientID:     "clientId",
+//			...
+//		},
+//		ClientSecretFile: "/path/to/client/secret",
+//	}
+type clientCredentialsConfig struct {
```

Comment:
Can't clientcredentials.Config be made to implement TokenSource instead of having a separate struct for this? I feel like this is leaking some concerns the caller shouldn't have.

Reply:
I'm not sure I get what you're saying. The API is for the user (here, the extension) to define a "Config". Here's my take:

* We need a new Config object to include extra options (`ClientIDFile`, `ClientSecretFile`)
* We need a TokenSource object on top of our config to essentially produce a proper Token by coming up with a proper `clientcredentials.Config`

Does this cover your question? Otherwise, could you elaborate?

Reply:
As mentioned in the issue the API derived from the upstream Go module is
* Have a Config object with all the configurations needed
* Produce a TokenSource object from the Config (using a Config method)
* Generate Tokens from the TokenSource object (using a TokenSource method)